packages/squiggle-lang/src/runners/PoolRunner.ts

```diff
@@ -0,0 +1,51 @@
+import { BaseRunner, RunParams, RunResult } from "./BaseRunner.js";
+
+type RunnerPoolThread = {
+  runner: BaseRunner;
+  job: Promise<RunResult> | undefined;
+};
+
+export class RunnerPool {
+  private makeRunner: () => BaseRunner;
+
+  maxThreads: number;
+
+  threads: RunnerPoolThread[] = [];
+
+  constructor(params: { makeRunner: () => BaseRunner; maxThreads?: number }) {
+    this.makeRunner = params.makeRunner;
+    this.maxThreads = params.maxThreads ?? 1;
+  }
+
+  async run(params: RunParams): Promise<RunResult> {
+    const unusedThread = this.threads.find(
+      (thread): thread is RunnerPoolThread & { runner: BaseRunner } =>
+        Boolean(thread.runner && !thread.job)
+    );
+    if (unusedThread) {
+      unusedThread.job = unusedThread.runner.run(params);
+      const result = await unusedThread.job;
+      unusedThread.job = undefined;
+      return result;
+    }
+
+    if (this.threads.length < this.maxThreads) {
+      // lazily allocate a new runner
+      this.threads.push({ runner: this.makeRunner(), job: undefined });
+    } else {
+      // or wait for the next job to complete
+      await Promise.race(this.threads.map((runner) => runner.job));
+    }
+    return this.run(params); // let's try again
+  }
+}
+
+export class PoolRunner extends BaseRunner {
```

Comment:
So far, this is working surprisingly well, I never had to debug it in the entire 3 weeks I worked on this branch. (This runner is the first piece of code I did here).

But it would be nice to have some introspection into what's going on in the pool, how many workers are busy, somewhere in `StateGraphViewer` in components; TODO, for later.